src/mmds/src/token.rs

```diff
@@ -0,0 +1,377 @@
+    ReadFromEntropyPool(io::Error),
+}
+
+impl fmt::Display for TokenError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &*self {
+            TokenError::EntropyPool(err) => write!(f, "Failed to open entropy pool: {}.", err),
+            TokenError::InvalidTtlValue(value) => write!(
+                f,
+                "Invalid time to live value provided for token: {}. \
+                Please provide a value between {} and {}.",
+                value, MIN_TOKEN_TTL_SECONDS, MAX_TOKEN_TTL_SECONDS,
+            ),
+            TokenError::ReadFromEntropyPool(err) => {
+                write!(f, "Failed to read from entropy pool: {}.", err)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum TokenStoreError {
+    /// Failed to generate session token.
+    TokenGenerationError(TokenError),
+    /// Session tokens limit exceeded.
+    TokensLimitExceeded,
+}
+
+impl fmt::Display for TokenStoreError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &*self {
+            TokenStoreError::TokenGenerationError(err) => {
+                write!(f, "Failed to generate session token: {}", err)
+            }
+            TokenStoreError::TokensLimitExceeded => {
+                write!(f, "Tokens limit exceeded: {}.", MAX_SESSIONS)
+            }
+        }
+    }
+}
+
+/// Structure for information regarding token lifetime.
+#[derive(Clone)]
+struct TokenTTL {
+    ttl_duration: Duration,
+    start_time: Instant,
+}
+
+impl TokenTTL {
+    /// Create a new session Token for MMDS requests.
+    fn new(ttl_seconds: u32) -> Self {
+        TokenTTL {
+            ttl_duration: Duration::from_secs(u64::from(ttl_seconds)),
+            start_time: Instant::now(),
+        }
+    }
+
+    /// Specifies whether the token has expired.
+    fn has_expired(&self) -> bool {
+        self.start_time.elapsed() >= self.ttl_duration
+    }
+
+    /// Validate the token time to live against bounds.
+    fn check_ttl(ttl_seconds: u32) -> bool {
+        (MIN_TOKEN_TTL_SECONDS..=MAX_TOKEN_TTL_SECONDS).contains(&ttl_seconds)
+    }
+}
+
+/// Structure for the Token used by MMDSv2.
+#[derive(Clone)]
+struct Token {
+    value: String,
+    token_info: TokenTTL,
+}
+
+impl Token {
+    /// Create a new session Token for MMDS requests.
+    fn new(value: String, token_info: TokenTTL) -> Self {
+        Token { value, token_info }
+    }
+
+    /// Generate a new random token with the specified lifetime.
+    fn generate(ttl_seconds: u32) -> Result<Self, TokenError> {
+        // Validate token time to live.
+        if !TokenTTL::check_ttl(ttl_seconds) {
+            return Err(TokenError::InvalidTtlValue(ttl_seconds));
+        }
+
+        let mut random_bytes = [0u8; TOKEN_LEN];
+        let mut token: String = String::new();
+        let mut random_file =
+            File::open(Path::new(RANDOMNESS_POOL)).map_err(TokenError::EntropyPool)?;
+
+        while token.len() < TOKEN_LEN {
```

Comment:
The duration of this loop depends on random factors :). Why not use base64 encoding or "force x to be alphanumeric" ?